crates/epaint/src/text/text_layout_types.rs

```diff
@@ -772,11 +772,16 @@ impl Galley {
         self.rect.size()
     }
 
-    pub(crate) fn round_output_to_gui(&mut self) {
-        for placed_row in &mut self.rows {
-            // NOTE: we round the size, but not the position, because the position should be _pixel_ aligned.
-            let row = Arc::make_mut(&mut placed_row.row);
-            row.size = row.size.round_ui();
+    pub(crate) fn round_output_to_gui(&mut self, round_row_sizes: bool) {
+        // This function is called in `Galley::concat` but by the time it is called
+        // Row sizes have already been rounded once after initial paragraph layout,
+        // therefore rounding them again here only causes an unnecessary make_mut.
+        if round_row_sizes {
+            for placed_row in &mut self.rows {
+                // NOTE: we round the size, but not the position, because the position should be _pixel_ aligned.
+                let row = Arc::make_mut(&mut placed_row.row);
+                row.size = row.size.round_ui();
```

Comment:
A less brittle solution would be to only call `make_mut` if not already rounded:

```rs
// Optimization: only call `make_mut` if necessary (can cause a deep clone)
let rounded_size = placed_row.row.size.round_ui();
if placed_row.row.size != rounded_size {
    Arc::make_mut(&mut placed_row.row).size = rounded_size;
}
```